streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.rocksdb.RocksIterator;
+
+import java.nio.ByteBuffer;
+import java.util.Set;
+
+class RocksDBPrefixIterator extends RocksDbIterator {
```

Comment:
I actually do not completely understand why we need a specific iterator for the prefix scan. We could just as good extend `RocksDBRangeIterator` to consider or not consider the end result of the range. We can do that because those iterator implementations are internal and the public API does not care which iterator is used as long as it implements interface `KeyValueIterator` and the behavior is correct. Extending and re-using `RocksDBRangeIterator` would lead to less code to maintain. Note, I agree that we need the public method `prefixScan()`, but what the implementation uses internally is not relevant for the KIP as long as it is correct. Did I miss something that imposes the implementation for a separate iterator?

Reply:
Well the only reason i chose to add a separate iterator is that for prefix scan there is no end key which could be known upfront. We could pass a null end key but that's prohibited in RocksDBRangeIterator constructor and i don't think we should be changing that condition or we could pass in the same value or from and to and add a condition in makeNext() for prefix scan. i thought having a separate iterator might be cleaner. You have some other approach in mind?

Reply:
Could you not use the same technique as you use in the in-memory key-value state store and the cache?

```
        final Bytes from = Bytes.wrap(prefixKeySerializer.serialize(null, prefix));
        final Bytes to = Bytes.increment(from);
```

Reply:
I tested with that approach. It worked for all cases apart from the abcd/abce case that you mentioned below. So, we will need a way to signal to the Iterator that for the case of range iteration for prefixScan, it should ignore the last key. I suppose we will have to add another flag similar to the ones used in in memory key-value store.
BTW, the other reason why I created a separate prefix iterator was that there was an old unused prefix iterator, So, I thought of repurposing it :)

Reply:
hey @cadonna , did you get a chance to look at the comment?

Reply:
Sorry for the late reply but I was on holidays.

> I suppose we will have to add another flag similar to the ones used in in memory key-value store.

This is what I meant with "extend RocksDBRangeIterator". 

I think less code is easier maintainable. In addition, I think the code is easier to follow if cache, in-memory, and RocksDB use the same technique for the prefix scan unless we have a technique with better performance that is specific to one of those components. So I would be in favor of just using `RocksDBRangeIterator`.

Reply:
Thanks for the reply. Sure, I would make the relevant changes.

Reply:
Awesome!

Reply:
As discusses, I have extended RocksDBRangeIterator to include the toInclusive flag. 
1 point to note here is that with this change, RocksDBPrefixIterator is no longer needed so I have removed it and it's relevant test case. 
Also, the unit tests for prefix scan, i have a separate PR [9717](https://github.com/apache/kafka/pull/9717) where i am writing the unit tests for RangeIterator. I plan to add the prefix scan cases there. would that be ok?